### DISCUSSION TITLE
posts/Go/ratelimit/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#330] [Announcements]
https://www.liwenzhou.com/posts/Go/ratelimit/ 

 李文周的Blog gin 限流 漏桶 令牌桶 ratelimit

**Comment by j2501987804** [comment]
```go
package main

import (
	"fmt"
	"github.com/gin-gonic/gin"
	"github.com/juju/ratelimit"
	"net/http"
	"time"
)

func RateLimitMiddleware(fillInterval time.Duration, cap int64) gin.HandlerFunc {
	bucket := ratelimit.NewBucket(fillInterval, cap)
	// bucket := ratelimit.NewBucket(5 * time.Second, 200)
	return func(c *gin.Context) {
		i := bucket.TakeAvailable(1)
		if i > 0 {
			c.String(http.StatusOK, "rate limit ,,,", i)
			c.Abort()
			return
		}
		fmt.Println(i)
		c.Next()
	}
}

func main() {
	r := gin.New()
	fillInterval := 20 * time.Second
	r.GET("/", RateLimitMiddleware(fillInterval, 2), func(c *gin.Context) {
		c.String(http.StatusOK, "this is main")
	})
	r.Run(":8080")
}
```
我这个是代表创建一个容量为2的桶,并且每20秒往桶里放一枚令牌吗?好像测试结果不是那么回事

**Comment by Q1mi** [comment]
> @j2501987804
> ```go
> package main
> 
> import (
> 	"fmt"
> 	"github.com/gin-gonic/gin"
> 	"github.com/juju/ratelimit"
> 	"net/http"
> 	"time"
> )
> 
> func RateLimitMiddleware(fillInterval time.Duration, cap int64) gin.HandlerFunc {
> 	bucket := ratelimit.NewBucket(fillInterval, cap)
> 	// bucket := ratelimit.NewBucket(5 * time.Second, 200)
> 	return func(c *gin.Context) {
> 		i := bucket.TakeAvailable(1)
> 		if i > 0 {
> 			c.String(http.StatusOK, "rate limit ,,,", i)
> 			c.Abort()
> 			return
> 		}
> 		fmt.Println(i)
> 		c.Next()
> 	}
> }
> 
> func main() {
> 	r := gin.New()
> 	fillInterval := 20 * time.Second
> 	r.GET("/", RateLimitMiddleware(fillInterval, 2), func(c *gin.Context) {
> 		c.String(http.StatusOK, "this is main")
> 	})
> 	r.Run(":8080")
> }
> ```
> 我这个是代表创建一个容量为2的桶,并且每20秒往桶里放一枚令牌吗?好像测试结果不是那么回事

TakeAvailable 返回的是此次取出的token数量，如果返回0表示此时没有可用的token。可用看一下该函数的注释https://github.com/juju/ratelimit/blob/f60b32039441cd828005f82f3a54aafd00bc9882/ratelimit.go#L221

**Comment by alexanyang** [comment]
这其实意味着，这个桶可以被攻击占满，从而使得其他用户无法访问该请求，还有例子中的写法是每次请求都是一个新桶，感觉无法限制

**Comment by Q1mi** [comment]
> @alexanyang
> 这其实意味着，这个桶可以被攻击占满，从而使得其他用户无法访问该请求

这里的示例默认是针对正常的请求，对于恶意请求可以前置或在限流过程中添加针对用户或IP的黑名单策略。

**Comment by alexanyang** [comment]
> @Q1mi
> > @alexanyang
> > 这其实意味着，这个桶可以被攻击占满，从而使得其他用户无法访问该请求
> 
> 这里的示例默认是针对正常的请求，对于恶意请求可以前置或在限流过程中添加针对用户或IP的黑名单策略。

那么每次都是新对象会不会达不到限制效果呢？

**Comment by f-ishelly** [comment]
> @alexanyang
> > @Q1mi
> > > @alexanyang
> > > 这其实意味着，这个桶可以被攻击占满，从而使得其他用户无法访问该请求
> > 
> > 这里的示例默认是针对正常的请求，对于恶意请求可以前置或在限流过程中添加针对用户或IP的黑名单策略。
> 
> 那么每次都是新对象会不会达不到限制效果呢？

那是闭包，只在注册路由中间件的时候创建对象
